apiserver/crossmodel/crossmodel.go

```diff
@@ -149,30 +219,81 @@ func makeOfferParamsFromOffer(offer jujucrossmodel.ApplicationOffer) params.Appl
 func (api *API) FindApplicationOffers(filters params.OfferFilters) (params.FindApplicationOffersResults, error) {
 	var result params.FindApplicationOffersResults
 
-	offerFilters, err := makeOfferFilterFromParams(filters.Filters)
-	if err != nil {
-		return result, err
+	models := make(map[string]Model)
+	filtersPerModel := make(map[string][]jujucrossmodel.ApplicationOfferFilter)
+
+	// Group the filters per model and then query each model with the relevant filters
+	// for that model.
+	modelUUIDs := make(map[string]string)
+	for _, f := range filters.Filters {
+		// Default model is the current model.
+		modelUUID := api.backend.ModelUUID()
+		model, ok := models[modelUUID]
+		if !ok {
+			var err error
+			model, err = api.backend.Model()
+			if err != nil {
+				return result, common.ServerError(errors.Trace(err))
+			}
+		}
+		// If the filter contains a model name, look up the details.
+		if f.ModelName != "" {
+			if _, ok := modelUUIDs[f.ModelName]; !ok {
```

Comment:
It seems like the second time you see a model name, you'll skip this if block because the UUID is found, but then you don't set the UUID and model, so the code after the if will use the default model rather than the model with that name. Add a test for this case?